### PR TITLE
Fix torch.utils._traceback on Python 3.11

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #101116
* #101220
* __->__ #101277

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

### CODE REVIEW

File: torch/utils/_traceback.py [hunk 1]
```diff
@@ -61,6 +89,22 @@ def report_compile_source_on_error():
                 # just have less accurate name info
                 if hasattr(code, 'replace'):
```
Comment (albanD) [inline torch/utils/_traceback.py:90]: nit: we only support 3.8+ so this check could be removed.

**Review by albanD — APPROVED** [review]
Sound good!

**Review by albanD — APPROVED** [review]

**Review by Skylion007 — APPROVED** [review]

File: torch/utils/_traceback.py [hunk 2]
```diff
@@ -50,17 +55,53 @@ def report_compile_source_on_error():
...
                 # FrameType directly, so just make one with compile
                 frame = tb.tb_frame
                 code = compile('__inspect_currentframe()', f.name, 'eval')
-                # Python 3.8 only.  In earlier versions of Python
-                # just have less accurate name info
-                if hasattr(code, 'replace'):
-                    code = code.replace(co_name=frame.f_code.co_name)
+                code = code.replace(co_name=frame.f_code.co_name)
+                # Python 3.11 only
+                if hasattr(frame.f_code, 'co_linetable'):
```
Comment (Skylion007) [inline torch/utils/_traceback.py:90]: This probably needs to be a Python version check to make mypy happy?

**Pull request closed without merging**